ZSTD_compressSequencesAndLiterals

Cyan4973 opened the pull request:
This PR introduces a variant of `ZSTD_compressSequences()`,
named `ZSTD_compressSequencesAndLiterals()`,
which accepts a buffer of literals instead of the source buffer.

This is meant to match specific limitations present in some transcoder operations,
in which it's easy to present Sequences and Literals coming from an external generator,
but it would be more complex or costly to restore the corresponding source.
As an example, one such use case happens within the Linux kernel,
where the source comes as scattered pages which do not form a continuous buffer.

This variant is a bit faster on its own than regular `ZSTD_compressSequences()`, by almost ~+20%,
as measured below using `fullbench` on a `i7-9700k`: 

| source | parser | `compressSequences` | `compressSequencesAndLiterals` | difference |
| --- | --- | --- | --- | --- |
| calgary.tar | level 19 | 805 MB/s | 960 MB/s | +19% |
| enwik6 | level 4 | 761 MBs | 917 MB/s | +20 % |

The difference is pretty consistent, driven primarily by a reduction in overhead during the sequence transcoding operation.

In addition, the goal is to cumulate the speed benefits of this variant with potential other benefits on the caller side, by no longer requiring presentation of the original data.

That being said, `ZSTD_compressSequencesAndLiterals()` also features some limitations:
- Only supports Explicit Block Delimiter mode. While this could be extended later on, it's unclear if it would be useful.
- Does not support Frame Checksum, since it doesn't see the original data. Requesting this checksum will result in an error.
- Does not write the content size within the frame header
- Cannot generate uncompressed blocks, since it doesn't have the original data. While this could be partially mitigated in some favorable cases, at the cost of complexity, there are many more complex scenarios for which it cannot be _completely_ eliminated. Thus, if the compression process tries to generate an uncompressed block, it results in a compression error.

For this reason, it's best to keep this method for compression of "small" data (typically <= 128 KB), where data is typically stored uncompressed when it doesn't compress well enough. This setup corresponds to typical kernel use cases like `zram`, `zswap` or `BtrFS`.

Cyan4973 committed:
publish new symbol ZSTD_compressSequencesAndLiterals()

Cyan4973 committed:
created ZSTD_storeSeqOnly()

makes it possible to register a sequence without copying its literals.

Cyan4973 committed:
move Sequences definition to zstd_compress_internal.h

they should not be in common/zstd_internal.h,
since these definitions are not shared beyond lib/compress/.

Cyan4973 committed:
codemod: seqDef -> SeqDef

SeqDef is a type name, so it should start with a Capital letter.
It's an internal symbol, no impact on public API.

Cyan4973 committed:
codemod: seqStore_t -> SeqStore_t

same idea, SeqStore_t is a type name, it should start with a Capital letter.

Cyan4973 committed:
codemod: ZSTD_sequenceLength -> ZSTD_SequenceLength

Cyan4973 committed:
codemod: symbolEncodingType_e -> SymbolEncodingType_e

Cyan4973 committed:
codemod: ZSTD_defaultPolicy_e -> ZSTD_DefaultPolicy_e

Cyan4973 committed:
ZSTD_entropyCompressSeqStore_internal() can accept an externally defined literals buffer

Cyan4973 committed:
created ZSTD_entropyCompressSeqStore_wExtLitBuffer()

can receive externally defined buffer of literals

Cyan4973 committed:
codemod: ZSTD_sequenceFormat_e -> ZSTD_SequenceFormat_e

since it's a type name.

Note: in contrast with previous names, this one is on the Public API side.
So there is a #define, so that existing programs using ZSTD_sequenceFormat_e still work.

Cyan4973 committed:
codemod: ZSTD_sequenceCopier -> ZSTD_SequenceCopier_f

Cyan4973 committed:
minor: simplify ZSTD_selectSequenceCopier

Cyan4973 committed:
scope: ZSTD_copySequencesToSeqStore*() are private to ZSTD_compress.c

no need to publish them outside of this unit.

Cyan4973 committed:
fix proper type for .forceNonContiguous

Cyan4973 committed:
enable proper type

Cyan4973 committed:
codemod: ZSTD_sequencePosition -> ZSTD_SequencePosition

Cyan4973 committed:
codemod: ZSTD_buildSeqStore_e -> ZSTD_BuildSeqStore_e

Cyan4973 committed:
codemod: ZSTD_matchState_t -> ZSTD_MatchState_t

Cyan4973 committed:
codemod: repcodes_t -> Repcodes_t

Cyan4973 committed:
codemod: rawSeqStore_t -> RawSeqStore_t

Cyan4973 committed:
codemod: ZSTD_blockCompressor -> ZSTD_BlockCompressor_f

Cyan4973 committed:
doc: add mention of frame checksum incompatibility

for new prototype ZSTD_compressSequencesAndLiterals()

Cyan4973 committed:
fix minor artifact error in single_file_lib

Cyan4973 committed:
minor simplification

Cyan4973 committed:
codemod: ZSTD_paramSwitch_e -> ZSTD_ParamSwitch_e

Cyan4973 committed:
codemod: ZSTD_cParamMode_e -> ZSTD_CParamMode_e

Cyan4973 committed:
ZSTD_SequenceCopier_f no returns the nb of bytes consumed from input

which feels much more natural

Cyan4973 committed:
minor: more accurate variable scope

Cyan4973 committed:
produced ZSTD_compressSequencesAndLiterals() as a separate pipeline

only supports explicit delimiter mode, at least for the time being

Cyan4973 committed:
added unit tests to ZSTD_compressSequencesAndLiterals()

seems to work as expected,
correctly control that `litSize` and `srcSize` are exactly correct.

Cyan4973 committed:
prepare fullbench for a refactor

Cyan4973 committed:
refactor fullbench

to make it easier to add new scenarios

Cyan4973 committed:
minor: test reordering

Cyan4973 committed:
minor variable renaming

Cyan4973 committed:
fullbench: preparation functions are now in charge of allocating buffers

Cyan4973 committed:
add the compressSequences() benchmark scenario

Cyan4973 committed:
fullbench: switch default generator to lorem ipsum

which creates more "realistic" scenarios than former compressible noise.

The legacy data generator remains accessible,
it is triggered when requesting an explicit compressibility factor (-P#).

Cyan4973 committed:
fullbench: new scenario: compressSequencesAndLiterals()

Cyan4973 committed:
minor conversion warning fix

Cyan4973 committed:
minor optimization for ZSTD_compressSequencesAndLiterals()

does not need to track and update internal `litPtr`.
note: does not measurably impact performance.

Cyan4973 committed:
minor doc update

Cyan4973 committed:
minor optimization: only track seqPos->posInSrc when validateSequences is enabled

note: very minor saving, no performance impact

Cyan4973 committed:
optimization: instantiate specialized version without Sequence checking code

results in +4% compression speed,
thanks to removal of branches in the hot loop.

Cyan4973 committed:
updated documentation on validateSequence

Cyan4973 committed:
minor: cleaner function parameter repcodeResolution

Cyan4973 committed:
change advanced parameter name: ZSTD_c_repcodeResolution

and updated its documentation.
Note: older name ZSTD_c_searchForExternalRepcodes remains supported via #define

Cyan4973 committed:
ZSTD_compressSequencesAndLiterals() now supports multi-blocks frames.

Cyan4973 committed:
added tests

check that ZSTD_compressAndLiterals() also controls that the `srcSize` field is exact.

Cyan4973 committed:
update Visual Studio solutions

Cyan4973 committed:
removed fullbench-dll project from visual solutions

Cyan4973 committed:
fixed incorrect assert

Cyan4973 committed:
attempt to silence Visual Studio warning about fopen()

Cyan4973 committed:
change name to ZSTD_convertSequences*()

Cyan4973 committed:
added benchmark for ZSTD_convertBlockSequences_wBlockDelim()

Cyan4973 committed:
improved speed of the Sequences converter

Cyan4973 commented:
Updated implementation, with slightly different prototype (does no longer request `srcSize`) and improved compression speed.

Cyan4973 committed:
fixed minor conversion warning

Cyan4973 committed:
fixed minor error in one benchmark scenario

terrelln commented on lib/zstd.h:
```diff
@@ -1633,27 +1635,57 @@ ZSTDLIB_STATIC_API size_t ZSTD_mergeBlockDelimiters(ZSTD_Sequence* sequences, si
+ * - Does not write the content size in frame header
+ * - If any block is incompressible, will fail and return an error
+ * - @litSize must be == sum of all @.litLength fields in @inSeqs. Any discrepancy will generate an error.
+ * - the buffer @literals must be larger than @litSize by at least 8 bytes.
```
Can we take another parameter `litCapacity` here? This seems like an easy constraint to break, and it would be good to force users to pass in a value here and check it directly.

Cyan4973 replied:
parameter `litCapacity` added

terrelln commented on tests/fuzz/sequence_compression_api.c:
```diff
@@ -232,7 +232,7 @@ static size_t roundTripTest(void* result, size_t resultCapacity,
                             const void* src, size_t srcSize,
```
We should test `ZSTD_compressSequencesAndLiterals()` here as well. We could just test both `ZSTD_compressSequences()` and `ZSTD_compressSequencesAndLiterals()`

Cyan4973 replied:
Fuzzer test for `ZSTD_compressSequencesAndLiterals()` added,
it piggy - backs on `ZSTD_compressSequences()`,
and gets triggered automatically whenever a compatible set of parameters is detected.

terrelln reviewed (commented):
Reviewed everything except the actual implementation in `zstd_compress.c` and `zstd_compress_internal.h`.

terrelln commented on lib/compress/zstd_compress.c:
```diff
@@ -6918,11 +6980,13 @@ ZSTD_compressSequences_internal(ZSTD_CCtx* cctx,
-        compressedSeqsSize = ZSTD_entropyCompressSeqStore(&cctx->seqStore,
-                                &cctx->blockState.prevCBlock->entropy, &cctx->blockState.nextCBlock->entropy,
-                                &cctx->appliedParams,
+        compressedSeqsSize = ZSTD_entropyCompressSeqStore_wExtLitBuffer(
```
It looks like this is the same as `ZSTD_entropyCompressSeqStore()`, can we call that instead? Or am I missing something?

Cyan4973 replied:
`ZSTD_entropyCompressSeqStore()` is invoking `ZSTD_entropyCompressSeqStore_wExtLitBuffer()`,
so they both converge to the same implementation.

The difference is that `ZSTD_entropyCompressSeqStore()` forces the `literals` buffer to be the one reserved into `cctx`, while `_wExtLitBuffer()` let the caller select where the buffer of `literals` is.

terrelln replied:
@Cyan4973 I understand that, but in this case the literals are in the seqStore, right?

There are so many parameters that it is hard to read, but it looks like this could just invoke `ZSTD_entropyCompressSeqStore()` instead, because the literals are `cctx->seqStore.litStart, (size_t)(cctx->seqStore.lit - cctx->seqStore.litStart)`.

Cyan4973 replied:
Ah, you meant this specific instance.
And yes, you are right, it could be replaced by `ZSTD_entropyCompressSeqStore()`.

This modification was made early during development, when I was trying to have a common pipeline for the 2 variants. This is a left over from this period. Since I gave up this approach, we now have 2 separate pipelines, so it's possible to restore the call to `ZSTD_entropyCompressSeqStore()`, especially if it helps readability.

terrelln commented on lib/compress/zstd_compress.c:
```diff
@@ -7022,6 +7092,246 @@ size_t ZSTD_compressSequences(ZSTD_CCtx* cctx,
     return cSize;
 }
 
+/*
+ * Note: Sequence validation functionality has been disabled (removed).
+ * This is helpful to find back simplicity, leading to performance.
+ * It may be re-inserted later.
+ */
```
Do we error if sequence validation was requested?

Cyan4973 replied:
nope, at this point, it's just ignored.
It could be a good idea to add such check and error out clearly.

Cyan4973 replied:
check added

terrelln commented on lib/compress/zstd_compress.c:
```diff
@@ -7022,6 +7092,246 @@ size_t ZSTD_compressSequences(ZSTD_CCtx* cctx,
+    /* Special case: empty frame */
+    if ((nbSequences == 1) && (inSeqs[0].litLength == 0)) {
+        U32 const cBlockHeader24 = 1 /* last block */ + (((U32)bt_raw)<<1);
+        RETURN_ERROR_IF(dstCapacity<4, dstSize_tooSmall, "No room for empty frame block header");
+        MEM_writeLE32(op, cBlockHeader24);
```
We have `MEM_writeLE24` right? Can we use it here to support `dstCapacity == 3`?

Cyan4973 replied:
done

terrelln commented on lib/compress/zstd_compress.c:
```diff
@@ -7022,6 +7092,246 @@ size_t ZSTD_compressSequences(ZSTD_CCtx* cctx,
+
+        RETURN_ERROR_IF(dstCapacity < ZSTD_blockHeaderSize, dstSize_tooSmall, "not enough dstCapacity to write a new compressed block");
+
+        compressedSeqsSize = ZSTD_entropyCompressSeqStore_internal(
```
Do we need to call `ZSTD_entropyCompressSeqStore_internal()` or can we call `ZSTD_entropyCompressSeqStore_wExtLitBuffer()`?

Cyan4973 replied:
The initial code was invoking `ZSTD_entropyCompressSeqStore_wExtLitBuffer()`.
But because `ZSTD_entropyCompressSeqStore_wExtLitBuffer()` is essentially `ZSTD_entropyCompressSeqStore()`, just with a selectable `litBuffer`,
it adds a few constraints, which favor the creation of uncompressed blocks.
In the normal case, it's fine (an uncompressed block is preferable to a compressed one with only 1 byte of saving),
but in the case of `ZSTD_compressSequencesAndLiterals()`, it is not good, because this variant is unable to generate an uncompressed block (since it doesn't see the original data), and the only possible outcome is to generate an error.
So we prefer skipping these additional constraints, by invoking the next stage directly, `ZSTD_entropyCompressSeqStore_internal()`.

Cyan4973 committed:
ZSTD_compressSequencesAndLiterals requires srcSize as parameter

this makes it possible to adjust windowSize to its tightest.

Cyan4973 committed:
added a test for ZSTD_compressSequencesAndLiterals

checks that srcSize is present in the frame header
and bounds the window size.

Cyan4973 committed:
add dedicated error code for special case

ZSTD_compressSequencesAndLiterals() cannot produce an uncompressed block

Cyan4973 committed:
ensure that srcSize is controlled

Cyan4973 committed:
fixed minor error in preparation of one fullbench scenario

Cyan4973 committed:
add a check, to return an error if Sequence validation is enabled

since ZSTD_compressSequencesAndLiterals() doesn't support it.

Cyan4973 committed:
added parameter litCapacity

to ZSTD_compressSequencesAndLiterals()
to enforce the litCapacity >= litSize+8 condition.

Cyan4973 committed:
minor: use MEM_writeLE24()

so that an empty frame needs only 3 bytes of dstCapacity.

terrelln reviewed (commented):
Looks good once we add the fuzzer

Cyan4973 committed:
restore invocation of ZSTD_entropyCompressSeqStore()

in the ZSTD_compressSequences() pipeline

Cyan4973 committed:
added fuzzer test for compressSequencesAndLiterals()

piggy-backing onto existing compressSequences() fuzzer test

Cyan4973 committed:
add one valid test case

ZSTD_compressSequencesAndLiterals() may return a specific error code
when data to compress is non-compressible.

Cyan4973 committed:
fixed another invalid scenario

compressSequencesAndLiterals() doesn't support sequence validation

Cyan4973 commented:
All points addressed
